Add Maxcube unit tests

unaiur opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add unit tests to maxcube integration.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml

```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [x] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

unaiur committed:
Add tests for maxcube component

MartinHjelmare commented on tests/components/maxcube/test_binary_sensor.py:
```diff
@@ -0,0 +1,56 @@
+    """Test a successful setup without any device."""
+    entities = []
+    handler = mock_cube_handle([])
+    setup_platform(mock_hass(handler), None, lambda l: entities.extend(l))
```
Please don't interact with integration details in the tests. Use the core to start the tests and assert test state.

https://developers.home-assistant.io/docs/development_testing#writing-tests-for-integrations

unaiur replied:
I am still working on it. Hope to have it ready in a few more days

unaiur replied:
Reworked and rebased on top of https://github.com/home-assistant/core/pull/48047
Please, review and merge that CR first.

MartinHjelmare replied:
That PR is merged. :+1:

unaiur committed:
Simplify maxcube integration

Device objects returned by maxcube-api dependency are stable, so
we do not need to resolve from the device address every time.

Also, refactor and unify how maxcube integration sets temperature & mode.

MartinHjelmare commented:
Please rebase this PR.

MartinHjelmare commented on tests/components/maxcube/conftest.py:
```diff
@@ -0,0 +1,132 @@
+    }
+
+
+@pytest.fixture
+def hass():
+    """Mock hass core."""
+    hass = get_test_home_assistant()
+    yield hass
+    hass.stop()
+
+
```
We already have a default `hass` fixture.
```suggestion
```

unaiur replied:
Removed

MartinHjelmare commented on tests/components/maxcube/conftest.py:
```diff
@@ -0,0 +1,132 @@
+
+def updated_state(hass, entity_id):
+    """Call update entity service and return the updated state of the given entity."""
+    hass.services.call(
```
Move time forward instead and let the default polling update the entity.
https://github.com/home-assistant/core/blob/272dffc3843fbca32fbb4f92639c0a815753a888/tests/common.py#L363-L365

unaiur replied:
I did not know about this feature. Using it!

unaiur replied:
I have replaced it by:
```
        async_fire_time_changed(hass, datetime.utcnow() + timedelta(minutes=10))
        await hass.async_block_till_done()
        state = hass.states.get(ENTITY_ID)
```

But state does not get updated like it does when calling that method. What else should I do?

unaiur replied:
Ohh, I need to use now() instead of utcnow(). Done!

MartinHjelmare replied:
Please use our dt util as much as possible. There's a `utcnow` function there.

MartinHjelmare commented on tests/components/maxcube/conftest.py:
```diff
@@ -0,0 +1,132 @@
+        cube.devices = [thermostat, wallthermostat, windowshutter]
+        cube.room_by_id.return_value = room
+        cube.devices_by_room.return_value = [thermostat, wallthermostat, windowshutter]
+        setup_component(hass, HA_DOMAIN, {})
```
```suggestion
```

unaiur replied:
Removed

MartinHjelmare commented on tests/components/maxcube/conftest.py:
```diff
@@ -0,0 +1,132 @@
+
+
+@pytest.fixture
+def cube(hass, hass_config, room, thermostat, wallthermostat, windowshutter):
```
```suggestion
async def cube(hass, hass_config, room, thermostat, wallthermostat, windowshutter):
```

unaiur replied:
Changed

MartinHjelmare commented on tests/components/maxcube/conftest.py:
```diff
@@ -0,0 +1,132 @@
+        cube.room_by_id.return_value = room
+        cube.devices_by_room.return_value = [thermostat, wallthermostat, windowshutter]
+        setup_component(hass, HA_DOMAIN, {})
+        setup_component(hass, DOMAIN, hass_config)
+        hass.block_till_done()
```
```suggestion
        assert await async_setup_component(hass, DOMAIN, hass_config)
        await hass.async_block_till_done()
```

unaiur replied:
Changed

MartinHjelmare commented on tests/components/maxcube/test_maxcube_climate.py:
```diff
@@ -0,0 +1,273 @@
+VALVE_POSITION = "valve_position"
+
+
+def _set_hvac_mode(hass, entity_id: str, mode: str):
```
Please inline the service calls in the tests.

unaiur replied:
Okay

MartinHjelmare commented on tests/components/maxcube/test_maxcube_climate.py:
```diff
@@ -0,0 +1,273 @@
+
+
+def _set_hvac_mode(hass, entity_id: str, mode: str):
+    hass.services.call(
```
We should use the async api in the tests. All the tests should be coroutine functions.

unaiur replied:
Replaced by await hass.servicies.async_call(...)

MartinHjelmare commented on tests/components/maxcube/test_maxcube_climate.py:
```diff
@@ -0,0 +1,273 @@
+    )
+
+
+def test_setup_thermostat(hass, cube: MaxCube):
```
```suggestion
async def test_setup_thermostat(hass, cube: MaxCube):
```

unaiur replied:
Even tests that does not use await keyword? I think that lint complains about that.

unaiur replied:
Okay, lint does not complain. Changed all tests to async.

MartinHjelmare commented:
Please don't squash commits after review has started to make it easier for readers to track changes.

unaiur commented:
Sorry. Do not know about that policy. Won't do it again.

unaiur commented:
The right is to post new commits to the branch and then merge them before merging or are we keeping all those commits in git history?

MartinHjelmare commented:
We'll squash when we merge, so there's nothing to worry about for the author regarding this. Just avoid squashing after review has started. :+1:

unaiur committed:
Use homeassistant.util.utcnow to retrieve current time

MartinHjelmare commented on homeassistant/components/maxcube/binary_sensor.py:
```diff
@@ -11,13 +11,10 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
-        for device in cube.devices:
-            name = f"{cube.room_by_id(device.room_id).name} {device.name}"
-
+        for device in handler.cube.devices:
```
Can we move the changes to the integration to a separate PR and just do the tests in this PR? It's better to add the tests first so we see that the tests still pass when making changes in the integration.

unaiur replied:
Done

unaiur committed:
Revert "Simplify maxcube integration"

This reverts commit 84d231d5bdfda9b7744d371d3025986b637a6a8b.

unaiur committed:
Make test pass again after rolling back integration changes

MartinHjelmare reviewed (approved):
Great!
